zebra-chain/src/serialization/zcash_serialize.rs

```diff
@@ -29,16 +29,68 @@ pub trait ZcashSerialize: Sized {
     }
 }
 
+/// Serialize a `Vec` as a compactsize number of items, then the items. This is
+/// the most common format in Zcash.
+///
+/// See `zcash_serialize_external_count` for more details, and usage information.
 impl<T: ZcashSerialize> ZcashSerialize for Vec<T> {
     fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
         writer.write_compactsize(self.len() as u64)?;
-        for x in self {
-            x.zcash_serialize(&mut writer)?;
-        }
-        Ok(())
+        zcash_serialize_external_count(self, writer)
     }
 }
 
+/// Serialize a typed `Vec` **without** writing the number of items as a
+/// compactsize.
+///
+/// In Zcash, most arrays are stored as a compactsize, followed by that number
+/// of items of type `T`. But in `Transaction::V5`, some types are serialized as
+/// multiple arrays in different locations, with a single compactsize before the
+/// first array.
+///
+/// ## Usage
+///
+/// Use `zcash_serialize_external_count` when the array count is determined by
+/// other data, or a consensus rule.
+///
+/// Use `Vec::zcash_serialize` for data that contains compactsize count,
+/// followed by the data array.
+///
+/// For example, when a single count applies to multiple arrays:
+/// 1. Use `Vec::zcash_serialize` for the array that has a data count.
+/// 2. Use `zcash_serialize_external_count` for the arrays with no count in the
+///    data, passing the length of the first array.
+///
+/// This function has a `zcash_` prefix to alert the reader that the
+/// serialization in use is consensus-critical serialization, rather than
+/// some other kind of serialization.
+//
+// we specifically want to serialize `Vec`s here, rather than generic slices
+#[allow(clippy::ptr_arg)]
+pub fn zcash_serialize_external_count<W: io::Write, T: ZcashSerialize>(
+    vec: &Vec<T>,
+    mut writer: W,
+) -> Result<(), io::Error> {
+    for x in vec {
+        x.zcash_serialize(&mut writer)?;
+    }
+    Ok(())
+}
+
+/// Serialize a raw byte `Vec` **without** writing the number of items as a
+/// compactsize.
+///
+/// This is a convenience alias for `writer.write_all(&vec)`.
+//
+// we specifically want to serialize `Vec`s here, rather than generic slices
+#[allow(clippy::ptr_arg)]
+pub fn zcash_serialize_bytes_external_count<W: io::Write>(
```

Comment:
Related to the [above comment](https://github.com/ZcashFoundation/zebra/pull/2017/files#r614227186),  this seems generic enough to be `zcash_serialize_vec_bytes` or something agnostic of `external_count`?

Reply:
It isn't generic, it is designed for a very specific and rare use case.
https://github.com/ZcashFoundation/zebra/pull/2017#discussion_r614378518